### PR TITLE
[WIP, TF] replace keras dense by keras.layers.DenseEinsum

### LINKED ISSUE #6771: Exported TF Bert model is much slower than that exported from Google's Bert
I benchmarked the Bert model exported from Huggingface TF Bert code and observed that it is much slower than Google's TF2 Bert model(https://github.com/tensorflow/models/tree/master/official/nlp/bert) on TF Serving (GPU device). When batch_size=1, Huggingface Bert model can be more than 200% slower than Google's one. Here are my benchmark results:

**latency: (ms/batch) on 1 GPU-V100**
Batch Size | google official | HF_TF2 | relative change |
-- | -- | -- | -- 
1 | 6.7 | 21.3 | 238.10%
2 | 9.4 | 24.2 | 171.91% 
4 | 14.4 | 28.1 | 99.29% 
8 | 24.0 | 36.9 | 55.70% 
16 | 46.6 | 58.6 | 26.57%
64 | 171.5 | 171.4 | 0.35% 
128 | 338.5 | 324.5 | -3.71% 

I used the latest Tensorflow profiling tool (https://www.tensorflow.org/tfx/serving/tensorboard) to compare these two:
 
**Hugging-face Bert-Base FP32 Batch_size = 128**
Screen Shot 2020-06-05 at 8 20 34 PM
**Google Bert-Base FP32 Batch_size = 128**
image (1)

Comparing the event traces of HF's and google's, we see that before each *MatMul* operation on GPU, there are always some CPU processes running (wasn’t able to get more info about these CPU processes because the Profiling tool doesn’t provide meaningful labels or description for those CPU processes but it seems like they are doing some gather/reduce operations) and the GPU is inactive for a while. If you look at the trace of google official, you can find that all GPU ops are compact and there are no idle time for GPU. In parallel, only one CPU process keeps running and doesn’t cause GPU ops to wait. The short idle time will hurt the performance more when batch_size or model_size is small because short time is spent on GPU calculations and idle time is not ignorable anymore. See time comparison grouped by type below:

**Hugging-face Bert-Base FP32 Batch_size = 4**  
image (3)

**Google Bert-Base FP32 Batch_size = 4**
image (4)

Under small batch_size=4, Hugging-face model has 60% GPU idle time while that of google model is only 25%. This also explains the reason why for small batch_size, there is over 200% slow down and for large batch_size=128, the situation gets better. 

The *MatMul* op is triggered by tf.keras.layers.Dense(), which is widely used in Transformer encoder self-attention, intermediate layer and output layer. In comparison, Google's Bert uses DenseEinsum()([link](https://github.com/tensorflow/models/blob/1ebff962832ec8a599bd67ac8c43b9ae1b05c3fa/official/nlp/transformer/attention_layer.py#L59-L69)) to replace all usage of Dense layer. I personally modified Huggingface's TF Bert code base to use DenseEinsum(). And the slow down issue got solved:

**latency: (ms/batch)  on 1 GPU-V100**
Batch Size | google official | HF_TF2 | relative change | After Fixing | relative change
-- | -- | -- | -- | -- | --
1 | 6.7 | 21.3 | 238.10% | 6.60 | 4.76%
2 | 9.4 | 24.2 | 171.91% | 8.90 | 0.00%
4 | 14.4 | 28.1 | 99.29% | 13.40 | -4.96%
8 | 24.0 | 36.9 | 55.70% | 22.10 | -6.75%
16 | 46.6 | 58.6 | 26.57% | 43.20 | -6.70%
64 | 171.5 | 171.4 | 0.35% | 158.80 | -7.03%
128 | 338.5 | 324.5 | -3.71% | 313.10 | -7.09%

Profile after fixing:
image (4) copy

I noticed that multiple issues may be related to this (https://github.com/huggingface/transformers/issues/6264).

Do you plan to solve this issue? Thanks. @patrickvonplaten

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Fixes #6771.

This PR might speed up TF runtime on GPU.
Note that the change requires TF 2.3.0

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
As a comparision. When running this line on current `master`:
```
TF_CPP_MIN_LOG_LEVEL=3 python examples/benchmarking/run_benchmark_tf.py --models bert-base-uncased --no_memory --batch_sizes 1 --sequence_lengths 128 256 512
```

one gets the following results: 

```
====================       INFERENCE - SPEED - RESULT       ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length     Time in s   
--------------------------------------------------------------------------------
      bert-base-uncased              1              128            0.006     
      bert-base-uncased              1              256            0.009     
      bert-base-uncased              1              512            0.017     
--------------------------------------------------------------------------------

====================        ENVIRONMENT INFORMATION         ====================
- transformers_version: 3.0.2
- framework: TensorFlow
- eager_mode: False
- use_xla: False
- framework_version: 2.3.0
- python_version: 3.6.10
- system: Linux
- cpu: x86_64
- architecture: 64bit
- date: 2020-09-01
- time: 11:23:30.836691
- fp16: False
- use_multiprocessing: True
- only_pretrain_model: False
- cpu_ram_mb: 32088
- use_gpu: True
- num_gpus: 1
- gpu: TITAN RTX
- gpu_ram_mb: 24217
- gpu_power_watts: 280.0
- gpu_performance_state: 2
- use_tpu: False
```

for a TITAN RTX GPU.

When running the above line on this branch, one gets the following results:

```
====================       INFERENCE - SPEED - RESULT       ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length     Time in s   
--------------------------------------------------------------------------------
      bert-base-uncased              1              128            0.006     
      bert-base-uncased              1              256            0.008     
      bert-base-uncased              1              512            0.016     
--------------------------------------------------------------------------------

====================        ENVIRONMENT INFORMATION         ====================
- transformers_version: 3.0.2
- framework: TensorFlow
- eager_mode: False
- use_xla: False
- framework_version: 2.3.0
- python_version: 3.6.10
- system: Linux
- cpu: x86_64
- architecture: 64bit
- date: 2020-09-01
- time: 11:28:12.021389
- fp16: False
- use_multiprocessing: True
- only_pretrain_model: False
- cpu_ram_mb: 32088
- use_gpu: True
- num_gpus: 1
- gpu: TITAN RTX
- gpu_ram_mb: 24217
- gpu_power_watts: 280.0
- gpu_performance_state: 2
- use_tpu: False
```

So, I cannot see a real difference here :-/ @jlei2

**Comment by patrickvonplaten** [comment]
I will see whether the benchmark results are better on a GPU-V100 GPU. 
@jlei2 - could you post the code you used to benchmark HF Bert vs. Google Bert ? This would help a lot for reproducability.

File: src/transformers/modeling_tf_bert.py [hunk 1]
```diff
@@ -237,6 +237,16 @@ def _linear(self, inputs):
         return tf.reshape(logits, [batch_size, length, self.vocab_size])
 
 
+def DenseEinsum(output_size, kernel_initializer=None, name=None):
+    return tf.keras.layers.experimental.EinsumDense(
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_bert.py:241]: Maybe check to see the TF version here?  We should support tf 2.0+ for the lib's internals

**Review by LysandreJik — CHANGES_REQUESTED** [review]

**Comment by LysandreJik** [comment]
oops didn't see it was [WIP]

**Comment by jplu** [comment]
@patrickvonplaten can you update your code with the version given by @jlei2 [here](https://github.com/jlei2/transformers/pull/2) when you have time please. Thanks a lot!
